metaflow/plugins/secrets/secrets_decorator.py

```diff
@@ -0,0 +1,228 @@
+            )
+        options = secret_spec_dict.get("options", {})
+        if not isinstance(options, dict):
+            raise MetaflowException(
+                "Bad @secrets specification - 'option' must be a dict - found %s"
+                % type(options)
+            )
+        return SecretSpec(secrets_backend_type, secret_id=secret_id, options=options)
+
+
+def validate_env_vars_across_secrets(all_secrets_env_vars):
+    vars_injected_by = defaultdict(str)
+    for secret_spec, env_vars in all_secrets_env_vars:
+        for k, v in env_vars.items():
+            if k in vars_injected_by:
+                raise MetaflowException(
+                    "Secret '%s' will inject '%s' as env var, and it is also added by '%s'"
+                    % (secret_spec.secret_id, k, vars_injected_by[k])
+                )
+            vars_injected_by[k] = secret_spec.secret_id
+
+
+def validate_env_vars_vs_existing_env(all_secrets_env_vars):
+    for secret_spec, env_vars in all_secrets_env_vars:
+        for k, v in env_vars.items():
+            if k in os.environ:
+                print(json.dumps(dict(os.environ), indent=4))
+                raise MetaflowException(
+                    "Secret %s will inject %s as env var, but it already exists in env"
+                    % (secret_spec.secret_id, k)
+                )
+
+
+def validate_env_vars(env_vars):
+    for k, v in env_vars.items():
+        if type(k) != str:
+            raise ValueError("Found non-str key %s (%s)" % (str(k), type(k)))
+        if type(v) != str:
+            raise ValueError("Found non-str value %s (%s)" % (str(v), type(v)))
+        if not re.fullmatch("[a-zA-Z_][a-zA-Z0-9_]*", k):
+            raise ValueError("Found invalid env var name '%s'." % k)
+
+
+def get_secrets_backend_provider(secrets_backend_type):
+    try:
+        return SECRETS_BACKEND_PROVIDERS[secrets_backend_type]
+    except KeyError:
+        raise MetaflowException(
+            "Unknown secrets backend type %s" % secrets_backend_type
+        )
+
+
+class SecretsDecorator(StepDecorator):
+    """
+    Specifies secrets to be retrieved and injected as environment variables prior to the execution of a step.
+
+    Parameters
+    ----------
+    vars : List[Union[str,dict]]
+        List of secret specs, defining how the secrets are to be retrieved
+    """
+
+    name = "secrets"
+    defaults = {"vars": []}
+
+    # def runtime_step_cli(
+    #    self, cli_args, retry_count, max_user_code_retries, ubf_context
+    # ):
+    def task_pre_step(
+        self,
+        step_name,
+        task_datastore,
+        metadata,
+        run_id,
+        task_id,
+        flow,
+        graph,
+        retry_count,
+        max_user_code_retries,
+        ubf_context,
+        inputs,
+    ):
+        # List of pairs (secret_spec, env_vars_from_this_spec)
+        all_secrets_env_vars = []
+        secret_specs = []
+
+        for secret_spec_str_or_dict in self.attributes["vars"]:
```

Comment:
I think `@secrets(from=[...])` would have been awesome - though Python doesn't allow this.  Good thing is `vars` matches, `@environment(vars={})` - BUT `@secrets` is doing something a bit different.

We will roll with `@secrets(vars=[])`, unless we come up with something better quickly (`@secrets(of=...)`?).

Reply:
ok, so I would maybe use something like this instead of `vars`:
  - `sources`: I actually kind of like that one -- it indicates where the secrets are coming from through either a path (the "." separated one) or through a descriptor block (the dictionary)
  - `secrets`: that one may be a bit weird and doing `@secrets(secrets=[])` may look weird
  - we could not name it anything at all and just have something like `@secrets('secret1', 'secret2')`. This would require changing a few things in the way we process decorators but mentioning as a possibility. It would solve this whole thing and possibly be cleaner. The change shouldn't be that much -- we just need to check if the first argument of the decorator is a function and if not, do the same thing that we do when we just have kwargs. Then, we package all the args into a _args attribute in the dictionary and tada, we are off. This change would be fully backward compatible and would just enable decorators to now just take arguments as opposed to just kwargs.

You could alternatively also use something like:
  - `names` for simple string secret names and `descriptors` for the dictionary descriptors to split them out. That may be a bit more complicated and doesn't really buy much but throwing it out there.